conda/models/environment.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+"""EXPERIMENTAL Conda environment data model"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from logging import getLogger
+from typing import TYPE_CHECKING
+
+from ..exceptions import CondaValueError
+
+if TYPE_CHECKING:
+    from typing import Any
+
+    from .match_spec import MatchSpec
+    from .records import PackageRecord
+
+
+log = getLogger(__name__)
+
+
+@dataclass
+class Environment:
+    """
+    **Beta** While in beta, expect both major and minor changes across minor releases.
+
+    Data model for a conda environment.
+    """
+
+    #: Prefix the environment is installed into (required).
+    prefix: str
+
+    #: The platform this environment may be installed on (required)
+    platform: str
+
+    #: Environment level configuration, eg. channels, solver options, etc.
+    config: dict[str, Any] = field(default_factory=dict)
+
+    #: Map of other package types that conda can install. For example pypi packages.
+    external_packages: dict[str, list[Any]] = field(default_factory=dict)
+
+    #: The complete list of specs for the environment.
+    #: eg. after a solve, or from an explicit environemnt spec
+    explicit_packages: list[PackageRecord] = field(default_factory=list)
+
+    #: Environment name
+    name: str | None = None
+
+    #: User requested specs for this environment.
+    requested_packages: list[MatchSpec] = field(default_factory=list)
+
+    #: Environment variables to be applied to the environment.
+    variables: dict[str, str] = field(default_factory=dict)
+
+    def __post_init__(self):
```

Comment:
I think we need some extra validation here:

- `platform` must be one of the non-noarch [KNOWN_SUBDIRS](https://github.com/conda/conda/blob/eb3b48191ae0135dd16c9a5d7cb40131d312dc77/conda/base/constants.py#L79)
- if both `requested_packages` and `explicit_packages` are set, each `requested_packages` entry should match one (and only one) record in `explicit_packages`.
- `explicit_packages` cannot contain records for more than one package name. IOW, `len([record.name for record in explicit_packages]) == len(explicit_packages)`.

Reply:
> Currently, users are not able to install a mix of explicit packages and requested specs. I think this validation should be done in not this layer.

My suggestion here comes from the notion that in this mode `requested_specs` is just metadata for `explicit_specs`. i.e. the list of specs that will end up in the `history` file, and the values that will populate the `requested_spec` field in the corresponding `conda-meta/*.json` file for that record.